api/routes.py

```diff
@@ -126,9 +126,34 @@ def wrapped_function(*args, **kwargs):
         return resp
     return update_wrapper(wrapped_function, f)
 
+def allows_admin_web(f):
+    # Override Flask's default behavior and intercept the OPTIONS method for
+    # every request so CORS headers can be added.
+    f.required_methods = getattr(f, 'required_methods', set())
+    f.required_methods.add("OPTIONS")
+    f.provide_automatic_options = False
+
+    def wrapped_function(*args, **kwargs):
+        if request.method == "OPTIONS":
+            resp = app.make_default_options_response()
+        else:
+            resp = make_response(f(*args, **kwargs))
+
+        admin_web_domains = app.manager.admin_web_domains
+        if admin_web_domains:
+            options = get_cors_options(
+                app, dict(origins=admin_web_domains,
+                          supports_credentials=True)
+            )
+            set_cors_headers(resp, options)
+
+        return resp
+    return update_wrapper(wrapped_function, f)
+
 h = ErrorHandler(app, app.config['DEBUG'])
 @app.errorhandler(Exception)
 @allows_patron_web
+@allows_admin_web
```

Comment:
Are my decorators in the right place? I understand that they run in inverse order... generally I put them underneath the `@allows_patron_web` decorator or below the `@returns_problem_detail` decorator.

Reply:
I was about to say that for this particular decorator the order doesn't matter much since the decorator either does nothing, or it sets an HTTP header.

But this question made me realize that @allows_patron_web sets the same header, potentially overriding it! That's a problem. I looked at [set_cors_headers](https://github.com/corydolphin/flask-cors/blob/master/flask_cors/core.py#L233) and it looks like you can only call it once per request. I think the second call will do nothing.

You should be able to verify this by using `curl` to request a URL that uses both decorators, like `/`. It should also be possible to make the tests in test_routes.py make assertions about what the decorators did to a request. Try writing a test that calls one of these methods and looks at the response headers -- you should get a real Response object back and it should have been run through the decorators.

I think the solution is to define a third decorator, @allows_patron_and_admin_web, which combines both sets of domains and uses them to call set_cors_headers once. Whenever you would need to use both decorators, use that third one instead.

Reply:
You're right. I was able to write a test showing that my cors headers were being overwritten. I have a fix that I'll push up soon.